lib/resolve.go

```diff
@@ -0,0 +1,138 @@
+package lib
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/qri-io/dataset"
+	"github.com/qri-io/qri/base"
+	"github.com/qri-io/qri/base/dsfs"
+	"github.com/qri-io/qri/dsref"
+	"github.com/qri-io/qri/fsi"
+)
+
+// ParseAndResolveRef combines reference parsing and resolution
+func (inst *Instance) ParseAndResolveRef(ctx context.Context, refStr, source string) (dsref.Ref, string, error) {
+	ref, err := dsref.Parse(refStr)
+
+	// bad case references are allowed-but-warned for backwards compatibility
+	if errors.Is(err, dsref.ErrBadCaseName) {
+		log.Error(dsref.ErrBadCaseShouldRename)
+		err = nil
+	} else if err != nil {
+		return ref, "", fmt.Errorf("%q is not a valid dataset reference: %w", refStr, err)
+	}
+
+	resolvedSource, err := inst.ResolveReference(ctx, &ref, source)
+	return ref, resolvedSource, err
+}
+
+// ResolveReference finds the identifier & HEAD path for a dataset reference.
+// the source parameter determines which subsystems of Qri to use when
+func (inst *Instance) ResolveReference(ctx context.Context, ref *dsref.Ref, source string) (string, error) {
+	if inst == nil {
+		return "", dsref.ErrNotFound
+	}
+
+	// Handle the "me" convenience shortcut
+	if ref.Username == "me" {
+		// TODO (b5) - this should be reading from a better place, and erroring if
+		// a canonical profile cannot be found for whatever reason
+		ref.Username = inst.cfg.Profile.Peername
+	}
+
+	resolvers, err := inst.resolveSources(source)
+	if err != nil {
+		return "", err
+	}
+
+	for _, resolver := range resolvers {
+		resolvedSource, err := resolver.ResolveRef(ctx, ref)
+		if err != nil {
+			if errors.Is(err, dsref.ErrNotFound) {
+				continue
+			} else {
+				return "", err
+			}
+		}
+
+		return resolvedSource, nil
+	}
+
+	return "", dsref.ErrNotFound
+}
+
+func (inst *Instance) resolveSources(source string) ([]dsref.Resolver, error) {
+	switch source {
+	case "":
+		return []dsref.Resolver{
+			inst.dscache,
+			inst.repo,
+			dsref.ParallelResolver(
+				inst.logbook,
```

Comment:
disagree that it causes problems. Yes it's slow, but logbook is only real source of truth here.

In recent refactoring I think I've found a nicer way to "deal with the slow": only falling back to `logbook.ResolveRef` within repo implementaionts. I want to avoid updating the `refstore` code to use InitIDs, having repo fall back to logbook means it'll only get called if dscache doesn't exist.